pkg/controller/volume/attach_detach_controller.go

```diff
@@ -19,63 +19,109 @@ limitations under the License.
+	reconcilerLoopPeriod time.Duration = 100 * time.Millisecond
 
-	// internalNodeInformer is the shared node informer used to fetch and store
-	// node objects from the API server. It is shared with other controllers
-	// and therefore the node objects in its store should be treated as
-	// immutable.
-	internalNodeInformer framework.SharedInformer
+	// reconcilerMaxSafeToDetachDuration is the maximum ammount of time the
+	// attach detach controller will wait for a volume to be safely detached
+	// from its node. Once this time has expired, the controller will assume the
+	// node or kubelet are unresponsive and will detach the volume anyway.
+	reconcilerMaxSafeToDetachDuration time.Duration = 10 * time.Minute
+)
 
-	// selfCreatedNodeInformer is true if the internalNodeInformer was created
-	// during initialization, not passed in.
-	selfCreatedNodeInformer bool
+// AttachDetachController defines the operations supported by this controller.
+type AttachDetachController interface {
+	Run(stopCh <-chan struct{})
 }
 
 // NewAttachDetachController returns a new instance of AttachDetachController.
 func NewAttachDetachController(
 	kubeClient internalclientset.Interface,
 	podInformer framework.SharedInformer,
 	nodeInformer framework.SharedInformer,
-	resyncPeriod time.Duration) AttachDetachController {
+	pvcInformer framework.SharedInformer,
+	pvInformer framework.SharedInformer,
+	resyncPeriod time.Duration,
+	cloud cloudprovider.Interface,
+	plugins []volume.VolumePlugin) (AttachDetachController, error) {
 	selfCreatedPodInformer := false
 	selfCreatedNodeInformer := false
+	selfCreatedPVCInformer := false
+	selfCreatedPVInformer := false
+	// TODO: The default resyncPeriod for shared informers is 12 hours, this is
+	// unacceptable for the attach/detach controller. For example, if a pod is
+	// skipped because the node it is scheduled to didn't set its annotation in
+	// time, we don't want to have to wait 12hrs before processing the pod
+	// again.
+	// Luckily https://github.com/kubernetes/kubernetes/issues/23394 is being
+	// worked on and will split resync in to resync and relist. Once that
+	// happens the resync period can be set to something much faster (30
+	// seconds).
+	// If that issue is not resolved in time, then this controller will have to
+	// consider some unappealing alternate options: use a non-shared informer
+	// and set a faster resync period even if it causes relist, or requeue
+	// dropped pods so they are continuously processed until it is accepted or
+	// deleted (probably can't do this with sharedInformer), etc.
 	if podInformer == nil {
-		podInformer = informers.CreateSharedPodInformer(kubeClient, resyncPeriod)
+		podInformer = informers.CreateSharedPodIndexInformer(kubeClient, resyncPeriod)
 		selfCreatedPodInformer = true
 	}
 	if nodeInformer == nil {
 		nodeInformer = informers.CreateSharedNodeIndexInformer(kubeClient, resyncPeriod)
 		selfCreatedNodeInformer = true
 	}
+	if pvcInformer == nil {
```

Comment:
Agreed it's ugly. It would be passed in `nil` for testing, but we could move this responsibility to the test. Will remove it.